README.md
# Stubborn Fetch

Fetch wrapper with built in retry

---

[![styled with prettier](https://img.shields.io/badge/styled_with-prettier-ff69b4.svg)](https://github.com/prettier/prettier)
[![Build Status](https://travis-ci.org/Quiq/stubborn-fetch.svg?branch=master)](https://travis-ci.org/Quiq/stubborn-fetch)

## Installation

```
npm i --save stubborn-fetch
```

or

```
yarn add stubborn-fetch
```

---

## Usage

```js
import StubbornFetch from 'stubborn-fetch';

new StubbornFetchRequest('/some/url').send().then(
  response => {
    // handle standard fetch response
  },
  error => {
    // handle standard fetch error
  },
);
```

---

## Parameters

### `url` - Url of the request to be made

```js
string;
```

### `fetchRequest` (optional) - Additional request [options](https://github.github.io/fetch/#options)

```js
Object = {
  method: 'get',
};
```

### `options` (optional) - Additional options

```js
Object = {
  timingFunction: 'exponential',
  maxDelay: 60000,
  debug: false,
  retries: 3,
  minimumStatusCodeForRetry: 400,
  retryOnNetworkFailure: false,
};
```

#### `timingFunction` - A function of the (form retryCount : delay in ms) to determine how long to wait between retries.

```js
string = 'exponential';
```

#### `maxDelay` - The maximum delay in ms between requests (upper bound on `timingFunction`)

```js
number;
```

#### `totalRequestTimeLimit` (optional) - The time limit across all retries of this request, after which the request will fail.

```js
number;
```

#### `retries` - How many times to attempt a request.

```js
number;
```

#### `minimumStatusCodeForRetry` - The lowest HTTP status code for which we will retry a request.

```js
number;
```

#### `retryOnNetworkFailure` - Whether we should retry a request when it fails due to a network issue, i.e. we did not get any response from server.

```js
boolean;
```

#### `maxErrors` (optional) - The maximum global error count we will tolerate across ALL requests. After this is hit, NO future requests will be sent.

```js
number;
```

#### `onError` - A function that will be called when a request attempt fails.

```js
(error: StubbornFetchError) => void
```

#### `shouldRetry` (optional) - Called for determining whether a retry attempt should occur. Takes precedence over other retry-related options.

```js
(error: StubbornFetchError, retries: number) => boolean;
```

#### `logger` (optional) - A class or object conforming to the `Logging` interface which we'll use for logging out request information and events.

```js
Logging = console;
```

## License

MIT


Changelog.md
# Changelog

## master

## 0.0.3

* Update module output

## 0.0.2

* Update Entry Points

## 0.0.1

* Initial release


src/__tests__/index.test.js
/* global jest */
import StubbornFetchError from '../error';
import 'isomorphic-fetch';
import fetchMock from 'fetch-mock';
import get from 'lodash/get';
import StubbornFetchRequest from '../index';

const getAllPathsInObject = (o: Object, currentPath: string = ''): Array<string> => {
  if (typeof o === 'object') {
    return Object.keys(o)
      .map(key => getAllPathsInObject(o[key], `${currentPath}${currentPath ? '.' : ''}${key}`))
      .reduce((flatArr, pathItem) => flatArr.concat(pathItem), []);
  }
  return [currentPath];
};

const subsetIsEqual = (superset, subset) => {
  const subsetPaths = getAllPathsInObject(subset);
  // eslint-disable-next-line no-restricted-syntax
  for (const path of subsetPaths) {
    if (get(superset, path) !== get(subset, path)) {
      return false;
    }
  }
  return true;
};

expect.extend({
  /**
   * Asserts that the received value is an instance of StubbornFetch error whose properties compose a superset of the passed object.
   * That is, for each path in the passed object, the path's value in the StubbornFetchError instance must be equal to that in the passed object.
   * It's similar to expect.objectContaining, but disregards structure of the objects as long as the value of each path are equivalent.
   */
  toBeStubbornFetchErrorWith(received, argument) {
    if (!(received instanceof StubbornFetchError)) {
      return {
        message: () => `Expected ${received} to be instance of StubbornFetchError`,
        pass: false,
      };
    }

    // eslint-disable-next-line no-restricted-syntax
    for (const [key, value] of Object.entries(argument)) {
      if (!received[key] || !subsetIsEqual(received[key], value)) {
        return {
          message: () =>
            `Expected ${received}.${key} to contain as a subset \n` +
            ` ${this.utils.printExpected(value)}\n` +
            `Received:\n` +
            ` ${this.utils.printReceived(received[key])}`,
          pass: false,
        };
      }
    }

    return {
      message: () =>
        `expected ${received} not to be a StubbornFetchError instance with indicated properties`,
      pass: true,
    };
  },
  /**
   * Asserts that the received value is an instance of Response whose properties compose a superset of the passed object.
   * That is, for each path in the passed object, the path's value in the Response instance must be equal to that in the passed object.
   * It's similar to expect.objectContaining, but disregards structure of the objects as long as the value of each path are equivalent.
   */
  toBeResponseWith(received, argument) {
    if (!(received instanceof Response)) {
      return {
        message: () => `Expected ${received} to be instance of Response`,
        pass: false,
      };
    }

    // eslint-disable-next-line no-restricted-syntax
    for (const [key, value] of Object.entries(argument)) {
      if (!received[key] || !subsetIsEqual(received[key], value)) {
        return {
          message: () =>
            `Expected ${received}.${key} to contain as a subset \n` +
            ` ${this.utils.printExpected(value)}\n` +
            `Received:\n` +
            ` ${this.utils.printReceived(received[key])}`,
          pass: false,
        };
      }
    }

    return {
      message: () =>
        `expected ${received} not to be a StubbornFetchError instance with indicated properties`,
      pass: true,
    };
  },
});

describe('StubbornFetch Service', () => {
  let options;
  let fetchRequest;
  let _originalSetTimeoutFunv, _originalDateNowFunv;
  const rejection = jest.fn();

  beforeAll(() => {
    fetchMock.mock('*', url => {
      if (url.indexOf('--fail--') !== -1) {
        return {
          status: 500,
          body: {status: 500, message: 'Uh oh!'},
        };
      } else if (url.indexOf('--fail400--') !== -1) {
        return {
          status: 402,
          body: {status: 402, message: 'Uh oh!'},
        };
      } else if (url.indexOf('--fail401--') !== -1) {
        return {
          status: 401,
          body: {status: 401, message: 'Uh oh!'},
        };
      } else if (url.indexOf('--fail429--') !== -1) {
        return {
          status: 429,
          body: {status: 429, message: 'Uh oh!'},
          headers: {'Retry-After': '2'},
        };
      } else if (url.indexOf('--fail429LongTime--') !== -1) {
        return {
          status: 429,
          body: {status: 429, message: 'Uh oh!'},
          headers: {'Retry-After': '360'},
        };
      } else if (url.indexOf('--fail429NoRetryAfterHeader--') !== -1) {
        return {
          status: 429,
          body: {status: 429, message: 'Uh oh!'},
        };
      }
      return {
        status: 200,
        body: {hello: 'world'},
      };
    });
  });

  beforeEach(() => {
    options = {
      debug: false,
      retries: 0,
    };
    fetchRequest = {};
  });

  afterEach(() => {
    // Reset StubbornFetch global error count
    StubbornFetchRequest.globalErrorCount = 0;
    StubbornFetchRequest.rateLimitedUntil = null;
    StubbornFetchRequest.enable();
    fetchMock.reset();
  });

  afterAll(() => {
    fetchMock.restore();
  });

  describe('dispatching request to isomorphic-fetch', () => {
    it('sends request with given URL', async () => {
      expect.assertions(1);

      await new StubbornFetchRequest('sendsReqWithGivenURL', fetchRequest, options).send();
      expect(fetchMock.called('sendsReqWithGivenURL', 'get')).toBe(true);
    });

    it('applies extended request data to fetch call', async () => {
      expect.assertions(1);

      fetchRequest = {
        headers: {
          accept: 'application/json',
          'content-type': 'application/json',
          'X-Foo': 'bar',
        },
      };

      await new StubbornFetchRequest('sendsReqWithData', fetchRequest, options).send();

      expect(fetchMock.lastCall('sendsReqWithData', 'get')[1]).toEqual(
        expect.objectContaining(fetchRequest),
      );
    });

    it('uses correct method (POST)', async () => {
      expect.assertions(1);

      fetchRequest = {
        method: 'POST',
      };

      await new StubbornFetchRequest('sendsReqWithMethod', fetchRequest, options).send();

      expect(fetchMock.lastCall('sendsReqWithMethod', 'post')[1]).toEqual(
        expect.objectContaining(fetchRequest),
      );
    });
  });

  describe('Promise fulfillment', () => {
    beforeEach(() => {
      options.retries = 0;
    });

    it('rejects promise on failure', () => {
      expect.assertions(1);

      return expect(
        new StubbornFetchRequest('--fail--callsReject', fetchRequest, options).send(),
      ).rejects.toBeStubbornFetchErrorWith({
        type: StubbornFetchError.types.HTTP_ERROR,
        data: JSON.stringify({
          response: {
            status: 500,
            body: {status: 500, message: 'Uh oh!'},
          },
        }),
      });
    });

    it('resolves promise on success', () => {
      expect.assertions(1);

      return expect(
        new StubbornFetchRequest('callsResolve', fetchRequest, options).send(),
      ).resolves.toBeResponseWith({
        status: 200,
        body: JSON.stringify({hello: 'world'}),
      });
    });
  });

  describe('retry delay', () => {
    beforeAll(() => {
      _originalSetTimeoutFunv = window.setTimeout.bind({});
      _originalDateNowFunv = Date.now.bind({});
      window.setTimeout = jest.fn(setTimeout);

      // We override date so that no time passes throughout test; this makes things much easier to reason about.
      const d = Date.now();
      Date.now = jest.fn(() => d);
    });

    afterAll(() => {
      window.setTimeout = _originalSetTimeoutFunv;
      Date.now = _originalDateNowFunv;
    });

    beforeEach(() => {
      options.retries = 2;
      options.timingFunction = 'exponential';
    });

    afterEach(() => {
      window.setTimeout.mockClear();
    });

    it('delays based on given backoff ', async () => {
      expect.assertions(3);

      try {
        await new StubbornFetchRequest('--fail--backoffFunc', fetchRequest, options).send();
      } catch (e) {
        const {calls} = setTimeout.mock;
        expect(calls.find(c => c[1] === 1500)).toBeDefined();
        expect(fetchMock.calls('--fail--backoffFunc', 'get').length).toEqual(2);
        expect(e).toBeStubbornFetchErrorWith({
          type: 'HTTP',
          data: {
            response: {
              status: 500,
            },
          },
        });
      }
    });

    describe('429 handling', () => {
      beforeEach(() => {
        options.timingFunction = 'constant';
      });

      describe('when 429 "try again at" time doesn\'t go beyond total request time limit', () => {
        it('delays next retry for the length of the rate limit + 100ms, if longer than that given by backoff function', async () => {
          expect.assertions(3);

          try {
            await new StubbornFetchRequest('--fail429--rateLimit', fetchRequest, options).send();
          } catch (e) {
            const {calls} = setTimeout.mock;
            expect(calls.pop()[1]).toBe(2000 + 100);
            expect(fetchMock.calls('--fail429--rateLimit', 'get').length).toEqual(2);
            expect(e).toBeStubbornFetchErrorWith({
              type: 'HTTP',
              data: {
                response: {
                  status: 429,
                },
              },
            });
          }
        });
      });

      describe('when rate limit will not be up before total request time limit is elapsed', () => {
        it('rejects immediately', async () => {
          options.totalRequestTimeLimit = 2000;
          expect.assertions(2);

          try {
            await new StubbornFetchRequest(
              '--fail429LongTime--rateLimitImmediateFail',
              fetchRequest,
              options,
            ).send();
          } catch (e) {
            expect(
              fetchMock.calls('--fail429LongTime--rateLimitImmediateFail', 'get').length,
            ).toEqual(1);
            expect(e).toBeStubbornFetchErrorWith({
              type: StubbornFetchError.types.RATE_LIMITED,
            });
          }
        });
      });

      describe('when no retry-after header is returned', () => {
        it('follows normal backoff function', async () => {
          expect.assertions(3);

          try {
            await new StubbornFetchRequest(
              '--fail429NoRetryAfterHeader--',
              fetchRequest,
              options,
            ).send();
          } catch (e) {
            const {calls} = setTimeout.mock;
            expect(calls.pop()[1]).toBe(1000);
            expect(fetchMock.calls('--fail429NoRetryAfterHeader--', 'get').length).toEqual(2);
            expect(e).toBeStubbornFetchErrorWith({
              type: 'HTTP',
              data: {
                response: {
                  status: 429,
                },
              },
            });
          }
        });
      });
    });
  });

  describe('number of retries', () => {
    beforeAll(() => {
      _originalSetTimeoutFunv = window.setTimeout.bind({});
      window.setTimeout = jest.fn(setTimeout);
    });

    afterAll(() => {
      window.setTimeout = _originalSetTimeoutFunv;
    });

    beforeEach(() => {
      options.retries = 2;
      options.timingFunction = 'exponential';
    });

    it('retries up to the maximum retry option', async () => {
      expect.assertions(3);
      try {
        await new StubbornFetchRequest('--fail--maxRetry', fetchRequest, options).send();
      } catch (e) {
        const {calls} = setTimeout.mock;
        expect(calls.length).toEqual(options.retries);
        expect(fetchMock.calls('--fail--maxRetry', 'get').length).toBe(2);
        expect(e).toBeStubbornFetchErrorWith({
          type: 'HTTP',
          data: {
            response: {
              status: 500,
            },
          },
        });
      }
    });
  });

  describe('request timeout', () => {
    beforeEach(() => {
      options.retries = -1;
      options.totalRequestTimeLimit = 10;

      jest.useFakeTimers();
    });

    afterEach(() => {
      jest.useRealTimers();
    });

    it('triggers timeout callback after defined totalRequestTimeLimit', () => {
      const promise = new StubbornFetchRequest('--fail--timesOut', fetchRequest, options)
        .send()
        .then(
          () => {},
          e => {
            rejection(e);
            expect(rejection).toBeCalledWith(
              expect.objectContaining({type: StubbornFetchError.types.TIMEOUT}),
            );
          },
        );

      // TODO: When we update jest, use runTimersToTime(11) here
      jest.runAllTimers();

      return promise;
    });
  });

  describe('retrying if status is above minimum', () => {
    beforeEach(() => {
      options.retries = 2;
      options.minimumStatusCodeForRetry = 400;
    });

    it('retries if status code greater than minimum', () => {
      expect.assertions(2);
      return new StubbornFetchRequest('--fail400--retryAboveMin', fetchRequest, options)
        .send()
        .then(
          () => {},
          e => {
            expect(fetchMock.calls('--fail400--retryAboveMin', 'get').length).toEqual(2);
            expect(e).toBeStubbornFetchErrorWith({
              type: 'HTTP',
              data: {
                response: {
                  status: 402,
                },
              },
            });
          },
        );
    });
  });

  describe('not retrying if status is below minimum', () => {
    beforeEach(() => {
      options.retries = 2;
      options.minimumStatusCodeForRetry = 505;
    });

    it('does not retry if status code less than minimum', () => {
      expect.assertions(2);
      return new StubbornFetchRequest('--fail400--retryBelowMin', fetchRequest, options)
        .send()
        .then(
          () => {},
          e => {
            expect(fetchMock.calls('--fail400--retryBelowMin', 'get').length).toEqual(1);
            expect(e).toBeStubbornFetchErrorWith({
              type: 'HTTP',
              data: {
                response: {
                  status: 402,
                },
              },
            });
          },
        );
    });

    it('does not retry if 401 is received regardless of minimumStatusCodeForRetry', () => {
      expect.assertions(2);

      options.minimumStatusCodeForRetry = 400;

      return new StubbornFetchRequest('--fail401--retryUnauth', fetchRequest, options).send().then(
        () => {},
        e => {
          expect(fetchMock.calls('--fail401--retryUnauth', 'get').length).toEqual(1);
          expect(e).toBeStubbornFetchErrorWith({
            type: 'HTTP',
            data: {
              response: {
                status: 401,
              },
            },
          });
        },
      );
    });
  });

  describe('global error limit', () => {
    beforeEach(() => {
      options.minimumStatusCodeForRetry = 400;
      options.retries = 3;
      options.maxErrors = 2;
    });

    it('does not retry more than maxErrors times', async () => {
      expect.assertions(3);
      try {
        await new StubbornFetchRequest('--fail--globalErrorLimit', fetchRequest, options).send();
      } catch (e) {
        expect(fetchMock.calls('--fail--globalErrorLimit', 'get').length).toBe(2);
        expect(StubbornFetchRequest.globalErrorCount).toBe(2);
        expect(e).toBeStubbornFetchErrorWith({
          type: StubbornFetchError.types.MAX_ERROS_EXCEEDED,
          data: {
            errorLimit: 2,
          },
        });
      }
    });

    it('will not initiate any new requests once global error limit has been reached', async () => {
      expect.assertions(3);

      StubbornFetchRequest.globalErrorCount = 2;

      try {
        await new StubbornFetchRequest(
          '--fail--globalErrorLimitAlreadyHit',
          fetchRequest,
          options,
        ).send();
      } catch (e) {
        expect(fetchMock.calls('--fail--globalErrorLimitAlreadyHit', 'get').length).toBe(0);
        expect(StubbornFetchRequest.globalErrorCount).toBe(2);
        expect(e).toBeStubbornFetchErrorWith({
          type: StubbornFetchError.types.MAX_ERROS_EXCEEDED,
          data: {
            errorLimit: 2,
          },
        });
      }
    });
  });

  describe('disabling (for burn it down)', () => {
    it('allows no further requests through when disabled', async () => {
      expect.assertions(2);

      StubbornFetchRequest.disable();

      try {
        await new StubbornFetchRequest('testDisabling', fetchRequest, options).send();
      } catch (e) {
        expect(fetchMock.calls('testDisabling', 'get').length).toBe(0);
        expect(e).toBeStubbornFetchErrorWith({
          type: StubbornFetchError.types.STUBBORN_FETCH_DISABLED,
        });
      }
    });

    it('allows no further retries through when disabled', async () => {
      expect.assertions(2);

      options.onError = () => StubbornFetchRequest.disable();
      options.retries = 2;

      try {
        await new StubbornFetchRequest(
          '--fail--noRetriesAfterDisable',
          fetchRequest,
          options,
        ).send();
      } catch (e) {
        expect(fetchMock.calls('--fail--noRetriesAfterDisable', 'get').length).toBe(1);
        expect(e).toBeStubbornFetchErrorWith({
          type: StubbornFetchError.types.STUBBORN_FETCH_DISABLED,
        });
      }
    });
  });

  describe('error hooks', () => {
    beforeEach(() => {
      options.retries = 2;
      options.onError = jest.fn();
    });

    it('calls error handler once per retry', () => {
      return new StubbornFetchRequest('--fail--errorHook', fetchRequest, options).send().then(
        () => {},
        () => {
          expect(options.onError.mock.calls.length).toEqual(2);
        },
      );
    });
  });
});


src/error.js
export function StubbornFetchError(type: string, message: ?string, data: Object = {}) {
  this.type = type;
  this.message = message || type;
  this.stack = Error().stack;
  this.data = data;
}
StubbornFetchError.prototype = Object.create(Error.prototype);
StubbornFetchError.prototype.name = 'StubbornFetchError';
StubbornFetchError.prototype.constructor = StubbornFetchError;
StubbornFetchError.types = {
  TIMEOUT: 'Timeout',
  MAX_ERROS_EXCEEDED: 'Max_Errors_Exceeded',
  NETWORK_ERROR: 'Network',
  STUBBORN_FETCH_DISABLED: 'Stubborn_Fetch_Disabled',
  HTTP_ERROR: 'HTTP',
  RATE_LIMITED: 'Rate_Limited',
};

export const ErrorFactory = {
  TIMEOUT: () => new StubbornFetchError(StubbornFetchError.types.TIMEOUT),
  MAX_ERRORS_EXCEEDED: (errorLimit: number) =>
    new StubbornFetchError('Max_Errors_Exceeded', null, {errorLimit}),
  NETWORK_ERROR: (underlyingError: Error) =>
    new StubbornFetchError('Network', null, {underlyingError}),
  STUBBORN_FETCH_DISABLED: () => new StubbornFetchError('Stubborn_Fetch_Disabled'),
  HTTP_ERROR: (response: Response) => new StubbornFetchError('HTTP', null, {response}),
  RATE_LIMITED: () => new StubbornFetchError('Rate_Limited'),
};

export default StubbornFetchError;


src/stubbornFetch.js
// @flow

import 'isomorphic-fetch';
import {ErrorFactory, StubbornFetchError} from './error';
import type {Logging} from '../types';

const TimingFunctions = {
  exponential: c => (c ** 2 - 1) / 2 * 1000,
  constant: () => 1000,
};

/**
 * @property timingFunction - A function of the (form retryCount : delay in ms) to determine how long to wait between retries.
 * Default to exponential backoff.
 * @property maxDelay - The maximum delay in ms between requests (upper bound on `timingFunction`)
 * @property totalRequestTimeLimit - The time limit across all retries of this request, after which the request will fail.
 * @property retries - How many times to attempt a request.
 * @property minimumStatusCodeForRetry - The lowest HTTP status code for which we will retry a request.
 * @property retryOnNetworkFailure - Whether we should retry a request when it fails due to a network issue, i.e. we did not get any response from server.
 * @property maxErrors - The maximum global error count we will tolerate across ALL requests. After this is hit, NO future requests will be sent.
 * @property onError - A function that will be called when a request attempt fails.
 * @property shouldRetry - Called for determining whether a retry attempt should occur. Takes precedence over other retry-related options.
 * @property logger - A class or object conforming to the `Logging` interface which we'll use for logging out request information and events.
 * Defaults to `console`
 */
export type StubbornFetchRequestOptions = {
  timingFunction: string,
  maxDelay: number,
  totalRequestTimeLimit?: number,
  retries: number,
  minimumStatusCodeForRetry: number,
  retryOnNetworkFailure: boolean,
  maxErrors?: number,
  onError?: (error: StubbornFetchError) => void,
  shouldRetry?: (error: StubbornFetchError, retries: number) => boolean,
  logger?: Logging,
};

/**
 * A retry wrapper around the Fetch API.
 */
class StubbornFetchRequest {
  static globalErrorCount: number = 0;
  static enabled: boolean = true;

  static disable() {
    this.enabled = false;
  }

  static enable() {
    this.enabled = true;
  }

  static rateLimitedUntil: ?number = null;

  options: StubbornFetchRequestOptions = {
    timingFunction: 'exponential',
    maxDelay: 60000,
    debug: false,
    retries: 3,
    minimumStatusCodeForRetry: 400,
    retryOnNetworkFailure: false,
  };

  logger: Logging;

  url: string;
  fetchRequest: Object;
  reqId = new Date().getTime();
  startTime: number;

  attemptCount = 0;
  error: ?StubbornFetchError;
  requestTimer: ?TimeoutID;
  rejectImmediately: (error: StubbornFetchError) => void;

  constructor(url: string, fetchRequest: ?Object, optionOverrides: ?Object = {}) {
    this.options = Object.assign({}, this.options, optionOverrides);
    this.logger = this.options.logger || console;
    this.url = url;
    this.fetchRequest = fetchRequest || {};
    this.error =
      typeof this.options.maxErrors === 'number' &&
      StubbornFetchRequest.globalErrorCount >= this.options.maxErrors
        ? ErrorFactory.MAX_ERRORS_EXCEEDED(this.options.maxErrors)
        : null;
  }

  _log(level: 'error' | 'warn' | 'info' | 'log' | 'log', message: string, data: ?Object) {
    const log = this.logger[level] || this.logger.log;
    const method = (this.fetchRequest && this.fetchRequest.method) || 'get';
    const logString = `${message.toUpperCase()}: [${method.toUpperCase()} ${this.url}]`;
    if (data) {
      log(logString, data);
    } else {
      log(logString);
    }
  }

  _getSanitizedRequest(): Object {
    const req = Object.assign({}, this.fetchRequest);
    delete req.headers;
    return req;
  }

  _startRequestTimer() {
    if (this.options.totalRequestTimeLimit) {
      this.requestTimer = setTimeout(() => {
        this.error = ErrorFactory.TIMEOUT();
        this.rejectImmediately(this.error);
        this._log('warn', 'time limit reached', {request: this._getSanitizedRequest()});
      }, this.options.totalRequestTimeLimit);
    }
  }

  _canRetry(error: StubbornFetchError): boolean {
    // If this request has already permanently failed, we definitely cannot retry
    if (this.error) {
      return false;
    }

    // Defer to downstream shouldRetry function if one was provided
    if (this.options.shouldRetry) {
      return this.options.shouldRetry(error, this.attemptCount);
    }

    let errorIsRetryable: boolean;
    switch (error.type) {
      case 'Network':
        errorIsRetryable = this.options.retryOnNetworkFailure;
        break;
      case 'HTTP':
        switch (error.data.response.status) {
          case 401:
          case 422:
            errorIsRetryable = false;
            break;
          default:
            errorIsRetryable = error.data.response.status >= this.options.minimumStatusCodeForRetry;
        }
        break;
      default:
        errorIsRetryable = false;
        break;
    }

    return (
      errorIsRetryable && (this.options.retries === -1 || this.attemptCount < this.options.retries)
    );
  }

  _requestGuard() {
    // Is StubbornFetch disabled?
    if (!StubbornFetchRequest.enabled) {
      throw ErrorFactory.STUBBORN_FETCH_DISABLED();
    }

    // Has global error limit been reached?
    if (
      typeof this.options.maxErrors === 'number' &&
      StubbornFetchRequest.globalErrorCount > this.options.maxErrors
    ) {
      this.error = ErrorFactory.MAX_ERRORS_EXCEEDED(this.options.maxErrors);
    }

    // Has current request permanently failed already?
    if (this.error) {
      throw this.error;
    }
  }

  _delayIfNeeded(): Promise<*> {
    // NOTE: We pad the rate limit time by 100ms just to be safe and not get rate limited again
    // Since this can be negative, if `rateLimitUntil` time has already been passed, we take max w.r.t. 0
    const delayDueToRateLimiting = Math.max(
      StubbornFetchRequest.rateLimitedUntil
        ? StubbornFetchRequest.rateLimitedUntil - Date.now() + 100
        : 0,
      0,
    );

    // NOTE: We want whichever is longer, the rate limit time or the result of the backoff function.
    // NOTE: The delay due to rate limiting is not subject to clamping
    const delay = Math.max(
      delayDueToRateLimiting,
      // Clamp (0, number, maxDelay)
      Math.max(
        0,
        Math.min(
          TimingFunctions[this.options.timingFunction](this.attemptCount),
          this.options.maxDelay,
        ),
      ),
    );
    return new Promise(resolve => {
      if (delay > 0) {
        this._log('info', `delay retry for ${delay} ms`);
      }
      window.setTimeout(resolve, delay);
    });
  }

  _handleError(e: StubbornFetchError) {
    StubbornFetchRequest.globalErrorCount++;
    if (
      typeof this.options.maxErrors === 'number' &&
      StubbornFetchRequest.globalErrorCount >= this.options.maxErrors
    ) {
      this.error = ErrorFactory.MAX_ERRORS_EXCEEDED(this.options.maxErrors);
    }

    // Error-specific logic
    switch (e.data.response.status) {
      case 401:
        this._log('warn', '401 received', {response: e.data.response});
        break;
      case 429:
        this._log('warn', 'rate limited', {response: e.data.response});

        // Adjust next retry time if response headers give us some hints
        if (e.data.response.headers.get('Retry-After')) {
          StubbornFetchRequest.rateLimitedUntil =
            Date.now() + parseInt(e.data.response.headers.get('Retry-After'), 10) * 1000;
          // Does this push us beyond the time limit?
          if (
            this.options.totalRequestTimeLimit &&
            StubbornFetchRequest.rateLimitedUntil - this.startTime >
              this.options.totalRequestTimeLimit
          ) {
            this.error = ErrorFactory.RATE_LIMITED();
            this.rejectImmediately(this.error);
          }
        }
        break;
    }
  }

  async _doAttempt(): Promise<Response> {
    // Throw an error immediately if we need to...
    this._requestGuard();

    this.attemptCount++;

    // Backoff, man
    await this._delayIfNeeded();

    // Check for error again, in case some condition changed while we were delaying
    // For example, another request may have pushed over the totalRequestTimeLimit limit
    this._requestGuard();

    let response;
    try {
      response = await fetch(this.url, this.fetchRequest);
    } catch (networkError) {
      const e = ErrorFactory.NETWORK_ERROR(networkError);
      this._handleError(e);
      throw this.error || e;
    }

    // NOTE: fetch resolves if network request was successful, but doesn't consider status codes.

    /******** Request SUCCESS! **********/
    if (response.status < 400) {
      // Success--resolve promise and send response to caller
      return response;
    }

    /******** Request FAILURE :( **********/
    const e = ErrorFactory.HTTP_ERROR(response);
    this._handleError(e);

    // If there is already an error defined on this whole request,
    // such as a timeout or max errors exceeded, throw this as it's more informative than the HTTP response.
    throw this.error || e;
  }

  async _runRequestLoop() {
    try {
      return await this._doAttempt();
    } catch (e) {
      if (this.options.onError) {
        this.options.onError(e);
      }

      // Retry if we can
      if (this._canRetry(e)) {
        return this._runRequestLoop();
      }

      // If we can't, then throw error (reject)
      throw e;
    }
  }

  async send(): Promise<Response> {
    this.startTime = Date.now();
    this._startRequestTimer();

    return new Promise((resolve, reject) => {
      this.rejectImmediately = reject;

      this._runRequestLoop()
        .then((response: Response) => {
          this._log('info', 'success', {response});
          resolve(response);
        })
        .catch((error: StubbornFetchError) => {
          this._log('error', error.type, {error});
          reject(error);
        })
        // $FlowIssue - flow doesn't know about Promise.finally() yet
        .finally(() => {
          if (this.requestTimer) {
            clearTimeout(this.requestTimer);
          }
        });
    });
  }
}

export default StubbornFetchRequest;
